File: examples/java/08_llm_integration_google/src/main/java/com/example/llm/LLMClient.java
package com.example.llm;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.JsonNode;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.*;

/**
 * LLM Client for OpenAI and Anthropic APIs.
 * 
 * Handles chat completions and tool calling for both providers.
 * COMPLEXITY: ~180 lines
 */
public class LLMClient {
    
    private static final ObjectMapper mapper = new ObjectMapper();
    private final HttpClient httpClient;
    
    private final String provider;
    private final String model;
    private final String apiKey;
    private final boolean ready;
    
    public LLMClient(String provider, String model) {
        this.provider = provider.toLowerCase();
        this.httpClient = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(30))
            .build();
        
        // Initialize based on provider
        switch (this.provider) {
            case "openai" -> {
                this.apiKey = System.getenv("OPENAI_API_KEY");
                this.model = model != null ? model : "gpt-4";
                this.ready = this.apiKey != null;
                if (!ready) {
                    System.err.println("  ⚠ OPENAI_API_KEY not set");
                }
            }
            case "anthropic" -> {
                this.apiKey = System.getenv("ANTHROPIC_API_KEY");
                this.model = model != null ? model : "claude-3-sonnet-20240229";
                this.ready = this.apiKey != null;
                if (!ready) {
                    System.err.println("  ⚠ ANTHROPIC_API_KEY not set");
                }
            }
            default -> throw new IllegalArgumentException("Unknown provider: " + provider);
        }
    }
    
    public boolean isReady() {
        return ready;
    }
    
    public String getProvider() {
        return provider;
    }
    
    public String getModel() {
        return model;
    }
    
    public LLMResponse chat(List<Map<String, String>> messages, 
                            List<Map<String, Object>> tools) throws Exception {
        return switch (provider) {
            case "openai" -> openaiChat(messages, tools);
            case "anthropic" -> anthropicChat(messages, tools);
            default -> throw new IllegalStateException("Unknown provider: " + provider);
        };
    }
    
    private LLMResponse openaiChat(List<Map<String, String>> messages,
                                    List<Map<String, Object>> tools) throws Exception {
        ObjectNode body = mapper.createObjectNode();
        body.put("model", model);
        body.put("temperature", 0.7);
        body.put("max_tokens", 1000);
        
        // Add messages
        ArrayNode messagesArray = body.putArray("messages");
        for (Map<String, String> msg : messages) {
            ObjectNode msgNode = messagesArray.addObject();
            msgNode.put("role", msg.get("role"));
            msgNode.put("content", msg.get("content"));
        }
        
        // Add tools if provided
        if (tools != null && !tools.isEmpty()) {
            ArrayNode toolsArray = body.putArray("tools");
            for (Map<String, Object> tool : tools) {
                ObjectNode toolNode = toolsArray.addObject();
                toolNode.put("type", "function");
                ObjectNode functionNode = toolNode.putObject("function");
                functionNode.put("name", (String) tool.get("name"));
                functionNode.put("description", (String) tool.get("description"));
                functionNode.set("parameters", mapper.valueToTree(tool.get("parameters")));
            }
            body.put("tool_choice", "auto");
        }
        
        HttpRequest request = HttpRequest.newBuilder()
            .uri(URI.create("https://api.openai.com/v1/chat/completions"))
            .header("Content-Type", "application/json")
            .header("Authorization", "Bearer " + apiKey)
            .POST(HttpRequest.BodyPublishers.ofString(body.toString()))
            .build();
        
        HttpResponse<String> response = httpClient.send(request, 
            HttpResponse.BodyHandlers.ofString());
        
        if (response.statusCode() != 200) {
            throw new RuntimeException("OpenAI API error: " + response.body());
        }
        
        JsonNode json = mapper.readTree(response.body());
        JsonNode choice = json.get("choices").get(0);
        JsonNode message = choice.get("message");
        
        String content = message.path("content").asText("");
        
        // Check for tool calls
        List<ToolCall> toolCalls = null;
        if (message.has("tool_calls")) {
            toolCalls = new ArrayList<>();
            for (JsonNode tc : message.get("tool_calls")) {
                String name = tc.path("function").path("name").asText();
                String argsStr = tc.path("function").path("arguments").asText();
                Map<String, Object> args = mapper.readValue(argsStr, Map.class);
                String id = tc.path("id").asText();
                toolCalls.add(new ToolCall(name, args, id));
            }
        }
        
        return new LLMResponse(content, toolCalls);
    }
    
    private LLMResponse anthropicChat(List<Map<String, String>> messages,
                                       List<Map<String, Object>> tools) throws Exception {
        // Separate system message from chat messages
        String systemMessage = null;
        List<Map<String, String>> chatMessages = new ArrayList<>();
        
        for (Map<String, String> msg : messages) {
            if ("system".equals(msg.get("role"))) {
                systemMessage = msg.get("content");
            } else {
                Map<String, String> chatMsg = new HashMap<>();
                chatMsg.put("role", msg.get("role"));
                chatMsg.put("content", msg.get("content"));
                chatMessages.add(chatMsg);
            }
        }
        
        ObjectNode body = mapper.createObjectNode();
        body.put("model", model);
        body.put("max_tokens", 1000);
        body.put("temperature", 0.7);
        
        if (systemMessage != null) {
            body.put("system", systemMessage);
        }
        
        // Add messages
        ArrayNode messagesArray = body.putArray("messages");
        for (Map<String, String> msg : chatMessages) {
            ObjectNode msgNode = messagesArray.addObject();
            String role = msg.get("role");
            // Map 'assistant' to 'assistant' for Anthropic
            msgNode.put("role", role.equals("assistant") ? "assistant" : "user");
            msgNode.put("content", msg.get("content"));
        }
        
        // Add tools if provided (Anthropic format)
        if (tools != null && !tools.isEmpty()) {
            ArrayNode toolsArray = body.putArray("tools");
            for (Map<String, Object> tool : tools) {
                ObjectNode toolNode = toolsArray.addObject();
                toolNode.put("name", (String) tool.get("name"));
                toolNode.put("description", (String) tool.get("description"));
                toolNode.set("input_schema", mapper.valueToTree(tool.get("parameters")));
            }
        }
        
        HttpRequest request = HttpRequest.newBuilder()
            .uri(URI.create("https://api.anthropic.com/v1/messages"))
            .header("Content-Type", "application/json")
            .header("x-api-key", apiKey)
            .header("anthropic-version", "2023-06-01")
            .POST(HttpRequest.BodyPublishers.ofString(body.toString()))
            .build();
        
        HttpResponse<String> response = httpClient.send(request, 
            HttpResponse.BodyHandlers.ofString());
        
        if (response.statusCode() != 200) {
            throw new RuntimeException("Anthropic API error: " + response.body());
        }
        
        JsonNode json = mapper.readTree(response.body());
        
        StringBuilder content = new StringBuilder();
        List<ToolCall> toolCalls = null;
        
        for (JsonNode block : json.get("content")) {
            String type = block.path("type").asText();
            if ("text".equals(type)) {
                content.append(block.path("text").asText());
            } else if ("tool_use".equals(type)) {
                if (toolCalls == null) {
                    toolCalls = new ArrayList<>();
                }
                String name = block.path("name").asText();
                Map<String, Object> args = mapper.convertValue(block.path("input"), Map.class);
                String id = block.path("id").asText();
                toolCalls.add(new ToolCall(name, args, id));
            }
        }
        
        return new LLMResponse(content.toString(), toolCalls);
    }
}

File: examples/python/08_llm_integration_google/llm_client.py
"""
LLM Client - Google A2A SDK Implementation

Integrates with REAL OpenAI and Anthropic APIs.
"""

import os
import json
from typing import List, Dict, Any, Optional, AsyncGenerator

# Try to import LLM libraries
try:
    import openai
    OPENAI_AVAILABLE = True
except ImportError:
    OPENAI_AVAILABLE = False

try:
    import anthropic
    ANTHROPIC_AVAILABLE = True
except ImportError:
    ANTHROPIC_AVAILABLE = False


class LLMClient:
    """
    Client for LLM APIs (OpenAI and Anthropic).
    
    Supports:
    - Chat completions
    - Tool/function calling
    - Streaming responses
    """
    
    def __init__(self, provider: str = "openai", model: Optional[str] = None):
        """
        Initialize LLM client.
        
        Args:
            provider: "openai" or "anthropic"
            model: Model name (optional, uses default if not specified)
        """
        self.provider = provider.lower()
        
        if self.provider == "openai":
            if not OPENAI_AVAILABLE:
                raise ValueError("OpenAI package not installed. Run: pip install openai")
            
            api_key = os.getenv("OPENAI_API_KEY")
            if not api_key:
                raise ValueError("OPENAI_API_KEY environment variable required")
            
            self.client = openai.AsyncOpenAI(api_key=api_key)
            self.model = model or "gpt-4"
            
        elif self.provider == "anthropic":
            if not ANTHROPIC_AVAILABLE:
                raise ValueError("Anthropic package not installed. Run: pip install anthropic")
            
            api_key = os.getenv("ANTHROPIC_API_KEY")
            if not api_key:
                raise ValueError("ANTHROPIC_API_KEY environment variable required")
            
            self.client = anthropic.AsyncAnthropic(api_key=api_key)
            self.model = model or "claude-3-sonnet-20240229"
        else:
            raise ValueError(f"Unknown provider: {provider}. Use 'openai' or 'anthropic'")
    
    async def chat(
        self,
        messages: List[Dict[str, str]],
        tools: Optional[List[Dict]] = None,
        temperature: float = 0.7,
        stream: bool = False
    ) -> Dict[str, Any]:
        """
        Send chat completion request.
        
        Args:
            messages: List of message dicts with role and content
            tools: Optional tool definitions for function calling
            temperature: Sampling temperature (0.0 - 2.0)
            stream: Whether to stream response (not fully implemented)
            
        Returns:
            Response dict with content and optional tool_calls
        """
        if self.provider == "openai":
            return await self._openai_chat(messages, tools, temperature, stream)
        else:
            return await self._anthropic_chat(messages, tools, temperature, stream)
    
    async def _openai_chat(
        self,
        messages: List[Dict[str, str]],
        tools: Optional[List[Dict]],
        temperature: float,
        stream: bool
    ) -> Dict[str, Any]:
        """OpenAI chat completion implementation."""
        params = {
            "model": self.model,
            "messages": messages,
            "temperature": temperature,
            "stream": stream
        }
        
        if tools:
            params["tools"] = tools
            params["tool_choice"] = "auto"
        
        if stream:
            # Return async generator for streaming
            return await self._openai_stream(params)
        
        response = await self.client.chat.completions.create(**params)
        
        message = response.choices[0].message
        result = {
            "content": message.content or "",
            "role": "assistant"
        }
        
        # Check for tool calls
        if message.tool_calls:
            result["tool_calls"] = [
                {
                    "name": tc.function.name,
                    "arguments": json.loads(tc.function.arguments),
                    "id": tc.id
                }
                for tc in message.tool_calls
            ]
        
        return result
    
    async def _openai_stream(self, params: Dict) -> AsyncGenerator[str, None]:
        """Stream OpenAI response (generator)."""
        stream = await self.client.chat.completions.create(**params)
        
        async for chunk in stream:
            if chunk.choices[0].delta.content:
                yield chunk.choices[0].delta.content
    
    async def _anthropic_chat(
        self,
        messages: List[Dict[str, str]],
        tools: Optional[List[Dict]],
        temperature: float,
        stream: bool
    ) -> Dict[str, Any]:
        """Anthropic chat completion implementation."""
        # Separate system message from chat messages
        system_msg = None
        chat_messages = []
        
        for msg in messages:
            if msg["role"] == "system":
                system_msg = msg["content"]
            else:
                chat_messages.append({
                    "role": msg["role"],
                    "content": msg["content"]
                })
        
        params = {
            "model": self.model,
            "messages": chat_messages,
            "temperature": temperature,
            "max_tokens": 4096
        }
        
        if system_msg:
            params["system"] = system_msg
        
        if tools:
            # Convert OpenAI tool format to Anthropic format
            anthropic_tools = []
            for tool in tools:
                if tool.get("type") == "function":
                    func = tool["function"]
                    anthropic_tools.append({
                        "name": func["name"],
                        "description": func["description"],
                        "input_schema": func["parameters"]
                    })
            params["tools"] = anthropic_tools
        
        if stream:
            return await self._anthropic_stream(params)
        
        response = await self.client.messages.create(**params)
        
        result = {
            "content": "",
            "role": "assistant"
        }
        
        # Extract content from response
        for content in response.content:
            if content.type == "text":
                result["content"] += content.text
            elif content.type == "tool_use":
                if "tool_calls" not in result:
                    result["tool_calls"] = []
                result["tool_calls"].append({
                    "name": content.name,
                    "arguments": content.input,
                    "id": content.id
                })
        
        return result
    
    async def _anthropic_stream(self, params: Dict) -> AsyncGenerator[str, None]:
        """Stream Anthropic response (generator)."""
        async with self.client.messages.stream(**params) as stream:
            async for text in stream.text_stream:
                yield text
